// @ngx-lib/multiselect/esm2015/lib/helper-elements/helper-elements.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/helper-elements/helper-elements.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Output, Input, ChangeDetectionStrategy } from '@angular/core';
export class HelperElementsComponent {
    constructor() {
        this.multiple = false;
        this.disabled = false;
        this.selectAllClicked = new EventEmitter();
        this.selectNoneClicked = new EventEmitter();
        this.resetClicked = new EventEmitter();
    }
    /**
     * @return {?}
     */
    selectAll() {
        this.selectAllClicked.emit();
    }
    /**
     * @return {?}
     */
    selectNone() {
        this.selectNoneClicked.emit();
    }
    /**
     * @return {?}
     */
    reset() {
        this.resetClicked.emit();
    }
}
HelperElementsComponent.decorators = [
    { type: Component, args: [{
                selector: 'ms-helper-elements',
                template: "<div class=\"helper-buttons\">\n  <fieldset [disabled]=\"disabled\">\n    <ng-container *ngIf=\"multiple\">\n      <button class=\"helper-button\" (click)=\"selectAll()\">\u2713 Select All</button>\n      <button class=\"helper-button\" (click)=\"selectNone()\">\u00D7 Select None</button>\n    </ng-container>\n    <button class=\"helper-button\" (click)=\"reset()\">\u21B6 Reset</button>\n  </fieldset>\n</div>",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [".helper-button{-moz-user-select:none;-ms-user-select:none;-webkit-user-select:none;border:1px solid rgba(0,0,0,.12);border-radius:.25rem;font-size:.75rem;font-weight:400;height:30px;max-height:30px;user-select:none;white-space:nowrap}fieldset{border:none;padding:0}.helper-buttons{border-bottom:1px solid #ddd;padding:2px}.mat-multiselect :host .helper-button{background-color:transparent;color:#444}.mat-multiselect :host .helper-button:not(:last-child){margin:0 2px}.bs-multiselect :host .helper-button{-webkit-tap-highlight-color:transparent;box-sizing:border-box;cursor:pointer;outline:0;padding:.25em}.bs-multiselect .helper-button:not(:last-child){margin-right:4px}"]
            }] }
];
/** @nocollapse */
HelperElementsComponent.ctorParameters = () => [];
HelperElementsComponent.propDecorators = {
    multiple: [{ type: Input }],
    disabled: [{ type: Input }],
    selectAllClicked: [{ type: Output }],
    selectNoneClicked: [{ type: Output }],
    resetClicked: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    HelperElementsComponent.prototype.multiple;
    /** @type {?} */
    HelperElementsComponent.prototype.disabled;
    /** @type {?} */
    HelperElementsComponent.prototype.selectAllClicked;
    /** @type {?} */
    HelperElementsComponent.prototype.selectNoneClicked;
    /** @type {?} */
    HelperElementsComponent.prototype.resetClicked;
}
