// @ngx-lib/multiselect/esm2015/lib/services/multiselect.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/multiselect.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/** @type {?} */
const matchSelectors = [
    'matches',
    'webkitMatchesSelector',
    'mozMatchesSelector',
    'msMatchesSelector',
    'oMatchesSelector'
];
export class NgxMultiselectService {
    constructor() { }
    /**
     * @param {?} node
     * @param {?} selector
     * @return {?}
     */
    pseudoClassExist(node, selector) {
        /** @type {?} */
        const nativeMatches = node.matches || node.msMatchesSelector;
        try {
            return nativeMatches.call(node, selector);
        }
        catch (error) {
            return false;
        }
    }
    /**
     * @param {?} el
     * @param {?} selector
     * @return {?}
     */
    closest(el, selector) {
        /** @type {?} */
        let matchesFn;
        // find vendor prefix
        matchSelectors.some((/**
         * @param {?} fn
         * @return {?}
         */
        function (fn) {
            if (typeof document.body[fn] === 'function') {
                matchesFn = fn;
                return true;
            }
            return false;
        }));
        /** @type {?} */
        let parent;
        // traverse parents
        while (el) {
            parent = el.parentElement;
            if (parent && parent[matchesFn](selector)) {
                return parent;
            }
            el = parent;
        }
        return false;
    }
    // TODO: make this logic to work to find all descendant groups
    /**
     * @param {?} collection
     * @param {?} groupProperty
     * @param {?} groupName
     * @return {?}
     */
    collectAllDescendants(collection, groupProperty, groupName) {
        /** @type {?} */
        const allDescendants = collection.filter((/**
         * @param {?} item
         * @return {?}
         */
        item => item[groupProperty] == groupName));
        allDescendants.concat(collection.filter((/**
         * @param {?} item
         * @return {?}
         */
        item => item.parent == groupName)));
        return allDescendants;
    }
    /**
     * @private
     * @param {?} collection
     * @param {?} groupProperty
     * @param {?} groupName
     * @return {?}
     */
    allDescendantsAreTicked(collection, groupProperty, groupName) {
        /** @type {?} */
        const allDescendants = this.collectAllDescendants(collection, groupProperty, groupName);
        /** @type {?} */
        const allAreTicked = allDescendants.every((/**
         * @param {?} d
         * @return {?}
         */
        d => d.ticked));
        return allAreTicked;
    }
    /**
     * @private
     * @param {?} collection
     * @param {?} groupProperty
     * @param {?} groupName
     * @return {?}
     */
    allDescendantsAreDisabled(collection, groupProperty, groupName) {
        /** @type {?} */
        const allDescendants = this.collectAllDescendants(collection, groupProperty, groupName);
        /** @type {?} */
        const allAreDisabled = allDescendants.every((/**
         * @param {?} d
         * @return {?}
         */
        d => d.disabled));
        return allAreDisabled;
    }
    /**
     * @param {?} options
     * @param {?} groupByProperty
     * @return {?}
     */
    optionsGrouping(options, groupByProperty) {
        /** @type {?} */
        const getAllUniqueGroupByPropertyValue = this.findUnique(options.map((/**
         * @param {?} item
         * @return {?}
         */
        item => item[groupByProperty])));
        /** @type {?} */
        const result = getAllUniqueGroupByPropertyValue.map((/**
         * @param {?} group
         * @return {?}
         */
        group => {
            /** @type {?} */
            const groupedValues = options.filter((/**
             * @param {?} o
             * @return {?}
             */
            o => o[groupByProperty] === group));
            return {
                name: group,
                values: groupedValues,
                ticked: groupedValues.every((/**
                 * @param {?} o
                 * @return {?}
                 */
                o => o.ticked)),
                disabled: groupedValues.every((/**
                 * @param {?} o
                 * @return {?}
                 */
                o => o.disabled))
            };
        }));
        return result;
    }
    /**
     * @param {?} expression
     * @return {?}
     */
    findUnique(expression) {
        return [...Array.from(new Set(expression))];
    }
    /**
     * @param {?} options
     * @param {?} groupByProperty
     * @return {?}
     */
    virtualOptionsGroupingFlatten(options, groupByProperty) {
        /** @type {?} */
        const allParentGroupedValues = this.findUnique(options.filter((/**
         * @param {?} o
         * @return {?}
         */
        o => !o.parent)).map((/**
         * @param {?} item
         * @return {?}
         */
        item => item[groupByProperty])));
        /** @type {?} */
        const subGroupedValues = this.findUnique(options.filter((/**
         * @param {?} o
         * @return {?}
         */
        o => o.parent)).map((/**
         * @param {?} __0
         * @return {?}
         */
        ({ name, parent }) => ({ name, parent }))));
        /** @type {?} */
        let result = [];
        allParentGroupedValues.forEach((/**
         * @param {?} group
         * @return {?}
         */
        group => {
            result.push({
                name: group,
                isGroup: true,
                ticked: this.allDescendantsAreTicked(options, groupByProperty, group),
                disabled: this.allDescendantsAreDisabled(options, groupByProperty, group)
            });
            /** @type {?} */
            const groupedValues = options
                .filter((/**
             * @param {?} o
             * @return {?}
             */
            o => o[groupByProperty] === group && !o.parent))
                .map((/**
             * @param {?} v
             * @return {?}
             */
            v => (Object.assign(Object.assign({}, v), { depth: 1 }))));
            result = [...result].concat(groupedValues);
            /** @type {?} */
            const childGroupedValues = subGroupedValues.filter((/**
             * @param {?} s
             * @return {?}
             */
            (s) => s.parent === group));
            childGroupedValues.forEach((/**
             * @param {?} c
             * @return {?}
             */
            c => {
                result.push({ name: c, parent: group, isGroup: true });
                /** @type {?} */
                const values = options.filter((/**
                 * @param {?} o
                 * @return {?}
                 */
                o => o[groupByProperty] === c)).map((/**
                 * @param {?} v
                 * @return {?}
                 */
                v => (Object.assign(Object.assign({}, v), { depth: 2 }))));
                result.concat(values);
            }));
        }));
        return result;
    }
    /**
     * @param {?} json
     * @return {?}
     */
    mirrorObject(json) {
        /** @type {?} */
        const ret = {};
        for (var key in json) {
            ret[json[key]] = key;
        }
        return ret;
    }
    /**
     * @param {?} collection
     * @param {?} propertyMap
     * @param {?=} groupedProperty
     * @return {?}
     */
    mapDatasourceToFields(collection, propertyMap, groupedProperty) {
        /** @type {?} */
        let keys = Object.keys(propertyMap);
        return collection.map((/**
         * @param {?} item
         * @return {?}
         */
        (item) => {
            /** @type {?} */
            let obj = groupedProperty ? { [groupedProperty]: item[groupedProperty] } : {};
            keys.reduce((/**
             * @param {?} a
             * @param {?} b
             * @return {?}
             */
            (a, b) => {
                obj[b] = item[propertyMap[b]];
            }), obj);
            return obj;
        }));
    }
}
NgxMultiselectService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
NgxMultiselectService.ctorParameters = () => [];
/** @nocollapse */ NgxMultiselectService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NgxMultiselectService_Factory() { return new NgxMultiselectService(); }, token: NgxMultiselectService, providedIn: "root" });
