// ngx-multiselect/multiselect/multiselect.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ComponentLoaderFactory } from '../component-loader';
import { PositioningService } from '../positioning';
import { DropdownState } from './dropdown.state';
import { DropdownContainerComponent } from './dropdown-container.component';
import { MultiselectComponent } from './multiselect.component';
export var MultiselectModule = (function () {
    function MultiselectModule() {
    }
    MultiselectModule.forRoot = function () {
        return {
            ngModule: MultiselectModule,
            providers: [
                ComponentLoaderFactory,
                PositioningService,
                DropdownState
            ]
        };
    };
    ;
    MultiselectModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [
                        MultiselectComponent,
                        DropdownContainerComponent
                    ],
                    exports: [MultiselectComponent],
                    entryComponents: [DropdownContainerComponent]
                },] },
    ];
    /** @nocollapse */
    MultiselectModule.ctorParameters = function () { return []; };
    return MultiselectModule;
}());
