// @ngx-lib/multiselect/esm2015/lib/multiselect.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/multiselect.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ChangeDetectionStrategy, ElementRef, ContentChild, TemplateRef, Output, EventEmitter, ViewChild, HostBinding, HostListener } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { NgxMultiselectService } from './services/multiselect.service';
import { forwardRef } from '@angular/core';
import { FilterOptionsComponent } from './filter-options/filter-options.component';
/** @type {?} */
export const DEFAULT_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef((/**
     * @return {?}
     */
    () => NgxMultiselectComponent)),
    multi: true
};
export class NgxMultiselectComponent {
    /**
     * @param {?} elementRef
     * @param {?} multiselectService
     */
    constructor(elementRef, multiselectService) {
        this.elementRef = elementRef;
        this.multiselectService = multiselectService;
        this._theme = 'material';
        this._isOpen = false;
        this.operationPendingQueue = [];
        // public variables
        this._selectedOptions = null;
        this._defaultPropertyMap = {
            id: 'id',
            name: 'name',
            disabled: 'disabled'
        };
        this._defaultPropertyMapLength = Object.keys(this._defaultPropertyMap).length;
        this.matMultiselect = true;
        this.bsMultiselect = false;
        // Input bindings
        this.disabled = false;
        this.color = 'blue';
        this.showHelperElements = true;
        this.showSearchFilter = true;
        this.showMaxLabels = 3;
        // Output bindings
        this.onOpen = new EventEmitter();
        this.onClose = new EventEmitter();
        this.onItemClick = new EventEmitter();
        this.onGroupItemClick = new EventEmitter();
        this.onSelectAll = new EventEmitter();
        this.onSelectNone = new EventEmitter();
        this.onReset = new EventEmitter();
        this.onClear = new EventEmitter();
        this.onSearchChange = new EventEmitter();
        this.onChange = (/**
         * @param {?} _
         * @return {?}
         */
        (_) => { });
        this.onTouched = (/**
         * @return {?}
         */
        () => { });
        this.filterOptionsList = (/**
         * @param {?} val
         * @return {?}
         */
        (val) => {
            /** @type {?} */
            const optionsCopy = this.getOptionsCopy();
            /** @type {?} */
            let result = optionsCopy;
            if (val) {
                result = optionsCopy.filter((/**
                 * @param {?} i
                 * @return {?}
                 */
                i => i.name && i.name.toLowerCase().indexOf(val.toLowerCase()) !== -1));
            }
            this.setOptions(result);
            this.prepopulateOptions(this._selectedOptions);
        });
    }
    /**
     * @return {?}
     */
    get optionsTemplate() {
        return this._optionsTemplate;
    }
    /**
     * @param {?} template
     * @return {?}
     */
    set optionsTemplate(template) {
        this._optionsTemplate = template;
    }
    /**
     * @return {?}
     */
    get theme() {
        return this._theme;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set theme(val) {
        this._theme = val;
        this.matMultiselect = val === 'material';
        this.bsMultiselect = val === 'bootstrap';
    }
    // Input binding with getter / setter
    /**
     * @param {?} value
     * @return {?}
     */
    set isOpen(value) {
        this._isOpen = value;
        // onOpen and onClose event will be fired by isOpen setter
        if (value) {
            // list populate, based on empty value
            this.filterOptionsList('');
            this.onTouched();
            this.onOpen.emit();
        }
        else {
            this.onClose.emit();
        }
    }
    /**
     * @return {?}
     */
    get isOpen() {
        return this._isOpen;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set propertyMap(val) {
        this._defaultPropertyMap = Object.assign(Object.assign({}, this._defaultPropertyMap), val);
    }
    /**
     * @return {?}
     */
    get multiple() {
        return this._multiple;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set multiple(value) {
        this.viewToModel(value ? [] : null);
        this._multiple = value;
    }
    /**
     * @param {?} collection
     * @return {?}
     */
    set options(collection) {
        if (!collection)
            return;
        this._optionsCopy = this.multiselectService.mapDatasourceToFields(collection, this._defaultPropertyMap, this.groupedProperty);
        /** @type {?} */
        const options = this.getOptionsCopy();
        this.setOptions(options);
        // Sometimes binding can happens lazily, 
        // so `finishPendingOperations` helps to execute such operations
        if (this.isOperationPending())
            this.finishPendingOperations();
    }
    // Adding pending operation in queue
    /**
     * @param {?} item
     * @return {?}
     */
    addOperation(item) {
        this.operationPendingQueue.push(item);
    }
    // Poping pending operation from queue sequentially
    /**
     * @return {?}
     */
    popOperation() {
        return this.operationPendingQueue.pop();
    }
    /*
        In future this code is going to resides inside different Service,
        This pendingOperation feature is fine grained in future,
        and can be used for multiple purpose like model update, collection update, etc.
      */
    // Extracting and finishing all pending operation
    /**
     * @return {?}
     */
    finishPendingOperations() {
        /** @type {?} */
        const operation = this.popOperation();
        this.prepopulateOptions(operation);
    }
    // Check pending operation queue status
    /**
     * @return {?}
     */
    isOperationPending() {
        return this.operationPendingQueue.length;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set initialValue(value) {
        this._initialValue = value;
    }
    /**
     * @return {?}
     */
    get initialValue() {
        return this._initialValue;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        // Set selected value for initial load of value
        if (value) {
            this.initialValue = value;
            this._options ? this.prepopulateOptions(value) : this.addOperation(value);
            this.formatPrepopulatedValues(value);
        }
    }
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    formatPrepopulatedValues(value) {
        /** @type {?} */
        let options = value;
        // TODO: can we improve below logic?
        if (Object.keys(this._defaultPropertyMap).length == this._defaultPropertyMapLength)
            return;
        /** @type {?} */
        const swappedPropertyMap = this.multiselectService.mirrorObject(this._defaultPropertyMap);
        if (this.multiple) {
            options.forEach((/**
             * @param {?} o
             * @return {?}
             */
            o => {
                o.id = o[swappedPropertyMap.id];
                o.name = o[swappedPropertyMap.name];
            }));
        }
        else {
            value.id = value[swappedPropertyMap.id];
            value.name = value[swappedPropertyMap.name];
            options = value;
        }
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    // All update to options should happen from below method.
    /**
     * @param {?} options
     * @return {?}
     */
    setOptions(options) {
        this._options = options;
    }
    /**
     * @return {?}
     */
    getOptions() {
        return this._options ? [...this._options] : [];
    }
    /**
     * @return {?}
     */
    getOptionsCopy() {
        return this._optionsCopy ? [...this._optionsCopy] : [];
    }
    /**
     * @return {?}
     */
    isValueSelected() {
        return this._selectedOptions && this._multiple ?
            this._selectedOptions.length :
            this._selectedOptions;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    searchChange(val) {
        this.filterOptionsList(val);
        this.onSearchChange.emit(val);
    }
    /**
     * @return {?}
     */
    close() {
        this.isOpen = false;
    }
    /**
     * @param {?} collection
     * @param {?} item
     * @return {?}
     */
    removeItem(collection, item) {
        item.ticked = false;
        /** @type {?} */
        const index = collection.findIndex((/**
         * @param {?} o
         * @return {?}
         */
        o => o.id === item.id));
        collection.splice(index, 1);
    }
    /**
     * @return {?}
     */
    toggleDropdown() {
        this.isOpen = !this.isOpen;
    }
    /**
     * @param {?} selected
     * @return {?}
     */
    prepopulateOptions(selected) {
        /** @type {?} */
        let selectedIds = [];
        selectedIds = this._multiple ?
            (selected || []).map((/**
             * @param {?} i
             * @return {?}
             */
            i => i.id)) :
            selected ? [selected.id] : [];
        this.setOptions(this.getOptions()
            .map((/**
         * @param {?} o
         * @return {?}
         */
        o => (Object.assign(Object.assign({}, o), { ticked: selectedIds.indexOf(o.id) !== -1 })))));
        // TODO: do we really need this reassignment?
        this.viewToModel(selected);
    }
    /**
     * @param {?} option
     * @return {?}
     */
    select(option) {
        /** @type {?} */
        let selectedOptions;
        option.ticked = !option.ticked;
        // TODO: Refactor below logic
        if (this._multiple) {
            selectedOptions = [...this._selectedOptions];
            /** @type {?} */
            let selectedIds = selectedOptions.map((/**
             * @param {?} i
             * @return {?}
             */
            i => i.id));
            if (selectedIds.indexOf(option.id) === -1) {
                // if selected item not exist in collection, push it
                selectedOptions.push(option);
            }
            else {
                // if selected item exist in collection, post it
                this.removeItem(selectedOptions, option);
            }
            selectedIds = selectedOptions.map((/**
             * @param {?} i
             * @return {?}
             */
            i => i.id));
        }
        else {
            // TODO: find optimized way to do below
            /** @type {?} */
            let val = option && option.id;
            /** @type {?} */
            let changedOptions = this.getOptions()
                .map((/**
             * @param {?} o
             * @return {?}
             */
            o => (Object.assign(Object.assign({}, o), { ticked: o.id == val }))));
            selectedOptions = changedOptions.find((/**
             * @param {?} i
             * @return {?}
             */
            i => i.ticked));
            this.setOptions(changedOptions);
            this.close();
        }
        this.viewToModel(selectedOptions);
        this.onItemClick.emit(option);
    }
    /**
     * @return {?}
     */
    selectAll() {
        /** @type {?} */
        let allSelectedOptions = this.getOptions()
            .map((/**
         * @param {?} o
         * @return {?}
         */
        o => (Object.assign(Object.assign({}, o), { ticked: true }))));
        this.setOptions(allSelectedOptions);
        this.viewToModel(allSelectedOptions);
        this.onSelectAll.emit();
    }
    /**
     * @return {?}
     */
    selectNone() {
        /** @type {?} */
        const options = this.getOptions().map((/**
         * @param {?} o
         * @return {?}
         */
        o => (Object.assign(Object.assign({}, o), { ticked: false }))));
        this.setOptions(options);
        this.viewToModel([]);
        this.onSelectNone.emit();
    }
    /**
     * @return {?}
     */
    borderBottom() {
        return this._isOpen ? { borderBottom: `1px solid ${this.matMultiselect ? this.color : 'transperant'}` } : {};
    }
    //TODO: Optimized below logic, it can be done in lesser steps
    /**
     * @param {?} group
     * @return {?}
     */
    selectGroup(group) {
        const { ticked, values } = group;
        /** @type {?} */
        const options = this.getOptions();
        /** @type {?} */
        let selectedValues = [...this._selectedOptions];
        /** @type {?} */
        let selectedIds = selectedValues.map((/**
         * @param {?} s
         * @return {?}
         */
        s => s.id));
        /** @type {?} */
        const allGroupOptionIds = values.map((/**
         * @param {?} v
         * @return {?}
         */
        v => v.id));
        // Get all ticked options
        // concat with selected options
        selectedValues = ticked
            ? selectedValues.concat(values)
            : selectedValues.filter((/**
             * @param {?} o
             * @return {?}
             */
            o => allGroupOptionIds.indexOf(o.id) === -1));
        // Find unique out of them
        selectedIds = this.multiselectService.findUnique(selectedValues.map((/**
         * @param {?} item
         * @return {?}
         */
        item => item.id)));
        // build selectedOptions array again
        selectedValues = options.filter((/**
         * @param {?} o
         * @return {?}
         */
        o => selectedIds.indexOf(o.id) !== -1));
        this.viewToModel(selectedValues);
        this.onGroupItemClick.emit(group);
    }
    /**
     * @return {?}
     */
    reset() {
        this.viewToModel(this.initialValue);
        this.prepopulateOptions(this.initialValue);
        this.onReset.emit();
    }
    // Responsible for updating value from view to model
    /**
     * @param {?} selected
     * @return {?}
     */
    viewToModel(selected) {
        if (this._selectedOptions !== selected) {
            this._selectedOptions = selected;
            this.onChange(selected);
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    clear(event) {
        /** @type {?} */
        let changedOptions = this.getOptions().map((/**
         * @param {?} o
         * @return {?}
         */
        o => (Object.assign(Object.assign({}, o), { ticked: false }))));
        this.setOptions(changedOptions);
        // no value is selected so passing null
        this.viewToModel(null);
        this.onClear.emit();
        this.close();
        event.stopPropagation();
    }
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    setDisabledState(isDisabled) {
        this.disabled = isDisabled;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        // Check if value have not been assigned then default to true
        if (typeof this._multiple === 'undefined') {
            this.multiple = true;
        }
    }
    // TODO: Consider creating a directive for this.
    // TODO: Also convert below to be work for element specific
    /**
     * @param {?} event
     * @return {?}
     */
    clickOutSide(event) {
        if (this.isOpen &&
            this.elementRef.nativeElement !== event &&
            !this.multiselectService.closest(event, 'ngx-multiselect')) {
            this.close();
        }
    }
}
NgxMultiselectComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngx-multiselect',
                template: "<div class=\"multiselect-wrapper\">\n  <button class=\"dropdown-button\" (click)=\"toggleDropdown()\" [ngClass]=\"{ open: isOpen }\">\n    <div class=\"selection\" [ngStyle]=\"borderBottom()\">\n      <!-- TODO: None selected template should be ng-template -->\n      <span class=\"none-selected\" *ngIf=\"!isValueSelected()\">None selected</span>\n      <span class=\"selected\" *ngIf=\"isValueSelected()\">\n        <span class=\"text\">{{_selectedOptions | displaySelectedValue: showMaxLabels}}</span>\n      </span>\n      <div class=\"symbols\">\n        <span class=\"close\" (click)=\"clear($event)\" *ngIf=\"isValueSelected() && !multiple\">\u00D7</span>\n        <span class=\"caret\"></span>\n      </div>\n    </div>\n  </button>\n  <div class=\"listing-wrapper\" *ngIf=\"isOpen\">\n    <ms-helper-elements\n      *ngIf=\"showHelperElements\"\n      [multiple]=\"multiple\"\n      [disabled]=\"disabled\"\n      (onClear)=\"onClear.emit()\"\n      (resetClicked)=\"reset()\"\n      (onSearchChange)=\"onSearchChange.emit($event)\"\n      (selectNoneClicked)=\"selectNone()\"\n      (selectAllClicked)=\"selectAll()\"\n      (selectNoneClicked)=\"selectNone()\"\n    >\n    </ms-helper-elements>\n    <ms-filter-options\n      *ngIf=\"showSearchFilter\"\n      #filterOptions\n      (onSearchChange)=\"searchChange($event)\"\n    ></ms-filter-options>\n    <ng-container *ngIf=\"!groupedProperty; else groupedTemplate\">\n      <ms-options\n        [optionsTemplate]=\"optionsTemplate\"\n        [options]=\"_options\"\n        [disabled]=\"disabled\"\n        (selectOption)=\"select($event)\"\n      ></ms-options>\n    </ng-container>\n  </div>\n</div>\n<ng-template #groupedTemplate>\n  <ms-grouped-options\n    [multiple]=\"multiple\"\n    [optionsTemplate]=\"optionsTemplate\"\n    [selectedOptions]=\"_selectedOptions\"\n    [options]=\"_options\"\n    [disabled]=\"disabled\"\n    [groupedProperty]=\"groupedProperty\"\n    (selectOption)=\"select($event)\"\n    (selectGroup)=\"selectGroup($event)\"\n  >\n  </ms-grouped-options>\n</ng-template>\n",
                providers: [DEFAULT_VALUE_ACCESSOR],
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host{display:inline-block;outline:none}.multiselect-wrapper{display:flex;flex-grow:1;width:240px}.selection{flex-grow:1;justify-content:space-between}.selection,.symbols{align-items:center;display:flex}.symbols{justify-content:flex-end}.symbols .caret{border-left:5px solid transparent;border-right:5px solid transparent;border-top:5px solid}.dropdown-button{-moz-user-select:none;-ms-user-select:none;-o-user-select:none;-webkit-user-select:none;background-color:#fff;border:none;cursor:pointer;display:flex;flex-grow:1;font-size:inherit;min-height:39px;padding:5px;user-select:none;white-space:normal}.listing-wrapper{background-color:#fff;border:1px solid rgba(0,0,0,.15);border-radius:4px;box-shadow:0 6px 12px rgba(0,0,0,.175);position:absolute;width:inherit;z-index:999}::ng-deep .options-container{height:200px;overflow-y:scroll}:host.mat-multiselect ::ng-deep .option{-webkit-tap-highlight-color:transparent;align-items:center;border-left:.5px solid #eee;display:flex;flex-flow:row;height:40px;max-height:40px;outline:0;overflow:hidden;padding:0 6px;text-overflow:ellipsis;white-space:nowrap}:host.mat-multiselect ::ng-deep .option.marked,:host.mat-multiselect ::ng-deep .option.marked:hover,:host.mat-multiselect ::ng-deep .option:hover:not(.message){background-image:linear-gradient(#e9e9e9,#f1f1f1);color:#555;cursor:pointer;justify-content:space-between}:host.mat-multiselect ::ng-deep .marked:after{content:\"\u2713\"}:host.mat-multiselect ::ng-deep .option.disabled,:host.mat-multiselect ::ng-deep .option.disabled:hover:not(.message),:host.mat-multiselect ::ng-deep .option.disabled label input:hover~span{color:#c4c4c4!important;cursor:not-allowed!important;pointer-events:none}:host.mat-multiselect ::ng-deep .none-selected{-webkit-text-fill-color:#7a7a7a;color:#7a7a7a;transition:none}:host.mat-multiselect ::ng-deep .dropdown-button{color:#000}:host.mat-multiselect ::ng-deep .selection{border-bottom:1px solid #000;color:#000;padding-bottom:10px}:host.mat-multiselect ::ng-deep .open .selection{border-bottom:1px solid #00f}:host.bs-multiselect ::ng-deep .marked{background-color:green}:host.bs-multiselect ::ng-deep .dropdown-button{background-image:linear-gradient(#fff,#f7f7f7);border:1px solid #c6c6c6;border-radius:4px;color:#555}:host.bs-multiselect ::ng-deep .option{align-items:center;background-color:transparent;border-radius:4px;box-shadow:inset 0 1px 1px rgba(0,0,0,.075);color:#555;display:flex;font-size:inherit;height:40px;max-height:40px;padding:0 6px}:host.bs-multiselect ::ng-deep .option:hover{background-image:linear-gradient(#c1c1c1,#999)!important;color:#fff!important;cursor:pointer}:host.bs-multiselect ::ng-deep .option.marked{background-image:linear-gradient(#e9e9e9,#f1f1f1);color:#555;cursor:pointer;display:flex;justify-content:space-between}:host.bs-multiselect ::ng-deep .marked:after{content:\"\u2713\";width:15px}:host.bs-multiselect ::ng-deep .option.disabled,:host.bs-multiselect ::ng-deep .option.disabled:hover,:host.bs-multiselect ::ng-deep .option.disabled label input:hover~span{color:#c4c4c4!important;cursor:not-allowed!important;pointer-events:none}"]
            }] }
];
/** @nocollapse */
NgxMultiselectComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: NgxMultiselectService }
];
NgxMultiselectComponent.propDecorators = {
    matMultiselect: [{ type: HostBinding, args: ['class.mat-multiselect',] }],
    bsMultiselect: [{ type: HostBinding, args: ['class.bs-multiselect',] }],
    disabled: [{ type: Input }],
    color: [{ type: Input }],
    groupedProperty: [{ type: Input }],
    showHelperElements: [{ type: Input }],
    showSearchFilter: [{ type: Input }],
    showMaxLabels: [{ type: Input }],
    _optionsTemplate: [{ type: ContentChild, args: [TemplateRef,] }],
    optionsTemplate: [{ type: Input }],
    theme: [{ type: Input }],
    isOpen: [{ type: Input }],
    propertyMap: [{ type: Input }],
    multiple: [{ type: Input }],
    options: [{ type: Input }],
    onOpen: [{ type: Output }],
    onClose: [{ type: Output }],
    onItemClick: [{ type: Output }],
    onGroupItemClick: [{ type: Output }],
    onSelectAll: [{ type: Output }],
    onSelectNone: [{ type: Output }],
    onReset: [{ type: Output }],
    onClear: [{ type: Output }],
    onSearchChange: [{ type: Output }],
    filterOptions: [{ type: ViewChild, args: ['filterOptions', { read: FilterOptionsComponent },] }],
    clickOutSide: [{ type: HostListener, args: ['document:click', ['$event.target'],] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgxMultiselectComponent.prototype._multiple;
    /**
     * @type {?}
     * @private
     */
    NgxMultiselectComponent.prototype._theme;
    /**
     * @type {?}
     * @private
     */
    NgxMultiselectComponent.prototype._optionsCopy;
    /**
     * @type {?}
     * @private
     */
    NgxMultiselectComponent.prototype._isOpen;
    /**
     * @type {?}
     * @private
     */
    NgxMultiselectComponent.prototype.operationPendingQueue;
    /** @type {?} */
    NgxMultiselectComponent.prototype._selectedOptions;
    /** @type {?} */
    NgxMultiselectComponent.prototype._defaultPropertyMap;
    /** @type {?} */
    NgxMultiselectComponent.prototype._defaultPropertyMapLength;
    /** @type {?} */
    NgxMultiselectComponent.prototype._options;
    /** @type {?} */
    NgxMultiselectComponent.prototype.matMultiselect;
    /** @type {?} */
    NgxMultiselectComponent.prototype.bsMultiselect;
    /** @type {?} */
    NgxMultiselectComponent.prototype.disabled;
    /** @type {?} */
    NgxMultiselectComponent.prototype.color;
    /** @type {?} */
    NgxMultiselectComponent.prototype.groupedProperty;
    /** @type {?} */
    NgxMultiselectComponent.prototype.showHelperElements;
    /** @type {?} */
    NgxMultiselectComponent.prototype.showSearchFilter;
    /** @type {?} */
    NgxMultiselectComponent.prototype.showMaxLabels;
    /** @type {?} */
    NgxMultiselectComponent.prototype._optionsTemplate;
    /** @type {?} */
    NgxMultiselectComponent.prototype.onOpen;
    /** @type {?} */
    NgxMultiselectComponent.prototype.onClose;
    /** @type {?} */
    NgxMultiselectComponent.prototype.onItemClick;
    /** @type {?} */
    NgxMultiselectComponent.prototype.onGroupItemClick;
    /** @type {?} */
    NgxMultiselectComponent.prototype.onSelectAll;
    /** @type {?} */
    NgxMultiselectComponent.prototype.onSelectNone;
    /** @type {?} */
    NgxMultiselectComponent.prototype.onReset;
    /** @type {?} */
    NgxMultiselectComponent.prototype.onClear;
    /** @type {?} */
    NgxMultiselectComponent.prototype.onSearchChange;
    /** @type {?} */
    NgxMultiselectComponent.prototype.filterOptions;
    /**
     * @type {?}
     * @private
     */
    NgxMultiselectComponent.prototype._initialValue;
    /** @type {?} */
    NgxMultiselectComponent.prototype.onChange;
    /** @type {?} */
    NgxMultiselectComponent.prototype.onTouched;
    /** @type {?} */
    NgxMultiselectComponent.prototype.filterOptionsList;
    /**
     * @type {?}
     * @private
     */
    NgxMultiselectComponent.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    NgxMultiselectComponent.prototype.multiselectService;
}
