// @ngx-lib/multiselect/esm2015/lib/filter-options/filter-options.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/filter-options/filter-options.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter, ChangeDetectionStrategy } from '@angular/core';
import { FormControl } from '@angular/forms';
export class FilterOptionsComponent {
    constructor() {
        this.onSearchChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    clearInputFilter() {
        this.filterName.setValue('');
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.filterName = new FormControl('');
        this.filterNameSubscription = this.filterName.valueChanges.subscribe((/**
         * @param {?} val
         * @return {?}
         */
        val => this.onSearchChange.emit(val)));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.filterNameSubscription.unsubscribe();
    }
}
FilterOptionsComponent.decorators = [
    { type: Component, args: [{
                selector: 'ms-filter-options',
                template: "<div class=\"input-search\">\n  <div class=\"field-infix\">\n    <input\n      autofocus\n      placeholder=\"Search options here..\"\n      type=\"text\"\n      [formControl]=\"filterName\"\n      ngDefaultControl\n      class=\"input-filter\"\n    />\n  </div>\n  <div class=\"field-suffix\">\n    <button type=\"button\" [disabled]=\"!filterName.value\" (click)=\"clearInputFilter()\" class=\"clear-button\">X</button>\n  </div>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [".input-search{align-items:stretch;display:flex;flex-grow:1;height:30px;padding:2px}.mat-multiselect :host ::ng-deep .input-search{border-bottom:1px solid rgba(0,0,0,.12)}.mat-multiselect :host ::ng-deep .input-filter{border:none;flex-grow:1;font:inherit;outline:none}.mat-multiselect :host ::ng-deep .field-infix{display:flex;flex:1;padding:5px}.mat-multiselect :host ::ng-deep .field-suffix{margin:3px;white-space:nowrap}.mat-multiselect :host ::ng-deep .field-suffix button{border:none;border-radius:10px;font-size:inherit;height:20px;min-width:20px;padding:0;width:20px}.bs-multiselect :host ::ng-deep .clear-button{background-color:#f1f1f1;border:none;border-radius:2px;color:#666;cursor:pointer;font-size:inherit;max-width:20px;min-width:20px;padding:0}.bs-multiselect :host ::ng-deep .input-filter{border:1px solid #ccc;border-radius:2px;box-shadow:inset 0 1px 1px rgba(0,0,0,.075);box-sizing:border-box;color:#666;cursor:pointer;font-size:inherit;height:30px;height:100%;padding-left:7px;width:100%}.bs-multiselect :host ::ng-deep .field-infix{display:flex;flex-grow:1}.bs-multiselect :host ::ng-deep .field-suffix{display:flex}"]
            }] }
];
/** @nocollapse */
FilterOptionsComponent.ctorParameters = () => [];
FilterOptionsComponent.propDecorators = {
    onSearchChange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    FilterOptionsComponent.prototype.filterName;
    /** @type {?} */
    FilterOptionsComponent.prototype.filterNameSubscription;
    /** @type {?} */
    FilterOptionsComponent.prototype.onSearchChange;
}
